src/Utils/TreeUtils.ts

```diff
@@ -0,0 +1,205 @@
+            sourceLogDialogIds: []
+        }
+    }
+
+    private addExtractorStep(parentNode: TreeNode, extractorStep: CLM.TrainExtractorStep, roundIndex: number, trainDialog: CLM.TrainDialog): TreeNode {
+        const child: TreeNode = {
+            name: "User",
+            userInput: extractorStep.textVariations.map(tv => { return { content: tv.text, trainDialogId: trainDialog.trainDialogId } }),
+            attributes: undefined,
+            children: [],
+            trainDialogIds: trainDialog.trainDialogId ? [trainDialog.trainDialogId] : [],
+            sourceLogDialogIds: trainDialog.sourceLogDialogId ? [trainDialog.sourceLogDialogId] : [],
+            roundIndex
+        }
+        parentNode.children.push(child)
+        return child
+    }
+
+    private toTreeScorerStep(scorerStep: CLM.TrainScorerStep): TreeScorerStep {
+        const entities: string[] = []
+        if (scorerStep.input.filledEntities && scorerStep.input.filledEntities.length > 0) {
+            scorerStep.input.filledEntities.forEach(fe => {
+                let entity = this.entities.find(e => e.entityId === fe.entityId)
+                if (entity) {
+                    entities.push(entity.entityName)
+                }
+                else {
+                    entities.push("UNKNOWN ENTITY")
+                }
+            })
+        }
+        return {
+            actionId: scorerStep.labelAction || "UNKNOWN ACTION",
+            memory: entities
+        }
+    }
+
+    private addScorerStep(parentNode: TreeNode, scorerStep: CLM.TrainScorerStep, roundIndex: number, scoreIndex: number, trainDialog: CLM.TrainDialog): TreeNode {
+        if (!scorerStep.labelAction) {
+            return parentNode
+        }
+
+        if (!parentNode.scorerSteps) {
+            parentNode.scorerSteps = []
+        }
+        parentNode.scorerSteps.push(this.toTreeScorerStep(scorerStep))
+        return parentNode
+    }
+
+    private addScorerSteps(parentNode: TreeNode, scorerSteps: CLM.TrainScorerStep[], roundIndex: number, trainDialog: CLM.TrainDialog): TreeNode {
+        let parent = parentNode
+        scorerSteps.forEach((scorerStep, scoreIndex) => {
+            parent = this.addScorerStep(parent, scorerStep, roundIndex, scoreIndex, trainDialog)
+        })
+        return parent
+    }
+
+    private doesRoundMatch(round1: TreeNode, round2: TreeNode): boolean {
+
+        // Check scorer steps array
+        if (round1.scorerSteps && !round2.scorerSteps ||
+            !round1.scorerSteps && round2.scorerSteps) {
+            return false
+        }
+        else if (!round1.scorerSteps && !round2.scorerSteps) {
+            return true
+        }
+        else {
+            return JSON.stringify(round1.scorerSteps) === JSON.stringify(round2.scorerSteps)
+        }
+    }
+
+    private findMatchingRound(parent: TreeNode, round: CLM.TrainRound, roundIndex: number, trainDialog: CLM.TrainDialog, testOnly: boolean): TreeNode | null {
```

Comment:
Hang on - why does this method new to build a temp tree, and why is the input parent mutated at line 185 ?

It feels like you should be able to navigate the existing tree without mutations to find a node containing a given round.

Reply:
badly named variable.  Throws it away if it finds a match of adds it to existing tree if not.  Renamed variable and added comment for clarification